.github/workflows/cd-circt.yml

```diff
@@ -1,11 +1,316 @@
+  # middle of the night isn't advantageous like typical nightly runs.
+  schedule:
+    - cron: '0 1 * * *'
 
 jobs:
-  hello-world:
-    name: Hello World
-    runs-on: ubuntu-20.04
+  # This job examines the CIRCT and Chisel repositories to determine the version
+  # of CIRCT that Chisel is using and the latest version of CIRCT that is
+  # available upstream.
+  #
+  # This job has to do a non-shallow checkout of both Chisel and CIRCT because
+  # it needs to examine tags.
+  determine-versions:
+    name: Determine Versions
+    runs-on: ubuntu-latest
+    outputs:
+      # The version of CIRCT that Chisel is _currently_ using.
+      version-current: ${{ steps.get-version-current.outputs.version }}
+      # The latest version of CIRCT that is available.
+      version-latest: ${{ steps.get-version-latest.outputs.version }}
+    steps:
+      - name: Clone chipsalliance/chisel
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          path: chisel
+      - name: Clone llvm/circt
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          repository: llvm/circt
+          path: circt
+      - name: Get Current Tag
+        id: get-version-current
+        run: |
+          cd chisel/
+          VERSION=$(jq -r '.version' < ${{ env.circt-config }})
+          echo version=$VERSION >> $GITHUB_OUTPUT
+          echo Current \`firtool\` version is: \`$VERSION\` >> $GITHUB_STEP_SUMMARY
+      - name: Get Latest Tag
+        id: get-version-latest
+        run: |
+          cd circt/
+          VERSION=$(git tag --list 'firtool-*' | tail -n1)
+          echo version=$VERSION >> $GITHUB_OUTPUT
+          echo Latest \`firtool\` version is: \`$VERSION\` >> $GITHUB_STEP_SUMMARY
+
+  create-pr:
+    # If a new version of CIRCT is available do the following logical actions.
+    #
+    #   1. Cherry-pick all commits from the `ci/ci-circt-nightly` staging branch
+    #      onto a new "compare" branch [1].
+    #   2. Create a new commit that updates the CIRCT version.
+    #   3. Create a new or update an existing Pull Request (PR) that merges this
+    #      branch with all changes into the `main` "base" branch [2] on GitHub.
+    #
+    # These logical steps are facilitated by a prerequisite step that assembles
+    # a bunch of JSON metadata about what is about to be done.  This JSON
+    # metadata is used to drive the cherry-pick action of (1), the commit
+    # message of (2), and the PR title and body of (3).
+    #
+    # This job re-clones both Chisel and CIRCT, in their entirety.  This is done
+    # because it is simpler to gate the JSON (and commit and PR messages) behind
+    # _this job_ where we know that a PR is going to be created.  We could save
+    # a CIRCT clone by doing this in the "Determine Versions" job at the cost of
+    # complicating this Workflow by needing to pass information between the
+    # jobs.
```

Comment:
FWIW, you always need to re-clone between different jobs, otherwise you'd have to be uploading and downloading the artifacts anyway. I'm not sure if this comment is really needed since it's basically inherent in how Github Actions work, see how every single job in `test.yml` has to clone Chisel.